New module: access Cisco ACI (network/aci/aci_rest)

dagwieers opened the pull request:
##### SUMMARY
This PR includes:
- Check-mode support
- Cosmetic changes to documentation
- Examples in YAML format
- Removal of incorrect requirements (for this module)
- Do not log passwords
- Implement native fetch_url instead of requests
- Use standard hostname, username and password parameters
- Add alias src for parameter config_file
- Add mutual exclusive content option for inline data (and show some inline examples)
- Add timeout parameter
- Add validate_certs parameter
- Handling ACI result output (identical for JSON as XML input)
- Parse/expose ACI error output to user

##### ISSUE TYPE
<!--- Pick one below and delete the rest: -->
 - New Module Pull Request

##### COMPONENT NAME
<!--- Name of the module/plugin/module/task -->
aci_rest

##### ANSIBLE VERSION
<!--- Paste verbatim output from “ansible --version” between quotes below -->
v2.4

dagwieers committed:
aci_rest: New module to access Cisco ACI

This PR includes:
- Relicense as GPLv3+
- Check-mode support
- Cosmetic changes to documentation
- Examples in YAML format
- Removal of incorrect requirements (for this module)
- Do not log passwords
- Implement native fetch_url instead of requests
- Use standard hostname, username and password parameters
- Add alias src for parameter config_file
- Add mutual exclusive content option for inline data (and show some inline examples)
- Add timeout parameter
- Add validate_certs parameter
- Handling ACI result output (identical for JSON as XML input)
- Parse/expose ACI error output to user

dagwieers commented:
@privateip This is the reworked aci_rest module, as good as it can be :-)

dagwieers commented:
@privateip This one is ready to be merged.

gundalow commented:
All new network modules in Ansible 2.4 (and above) must have either
* Unit tests
* integration tests that can be run via Shippable
http://docs.ansible.com/ansible/dev_guide/testing_integration.html#network-tests

dagwieers commented:
Luckily v2.4 is not released then ;-)

dagwieers committed:
Lower case method, add use_ssl, Use python dicts

This commit includes:
- Use lowercase method names
- Add `use_ssl` parameter (not the `protocol` parameter)
- Use a python dict for the request data (not a JSON string)
- Documentation improvements

dagwieers committed:
Ensure one of 'content' or 'src' is provided

dagwieers committed:
Fix issue with totalCount being a string in JSON

This fixes the problem with JSON output where totalCount is a string and
not an integer.

This fixes jedelman8/aci-ansible#7

dagwieers committed:
Improve code documentation

Commit:
Improve error handling and module response

dagwieers committed:
Small typo

dagwieers committed:
Improve documentation and examples

dagwieers committed:
Keep protocol parameter, but deprecate it

Commit:
Extrude aci functions from module_utils

dagwieers committed:
aci_rest: Add unit tests

dagwieers commented:
@mattclay Weird MacOSX integration test failure:
```
2017-07-17 00:28:32 TASK [expect : test chdir] *****************************************************
2017-07-17 00:28:32 task path: /private/var/root/ansible/test/integration/targets/expect/tasks/main.yml:62
2017-07-17 00:28:33 changed: [testhost] => {
2017-07-17 00:28:33     "changed": true, 
2017-07-17 00:28:33     "cmd": "pwd", 
2017-07-17 00:28:33     "delta": "0:00:00.707304", 
2017-07-17 00:28:33     "end": "2017-07-16 20:28:33.639495", 
2017-07-17 00:28:33     "failed": false, 
2017-07-17 00:28:33     "invocation": {
2017-07-17 00:28:33         "module_args": {
2017-07-17 00:28:33             "chdir": "/var/root/ansible_testing", 
2017-07-17 00:28:33             "command": "pwd", 
2017-07-17 00:28:33             "creates": null, 
2017-07-17 00:28:33             "echo": false, 
2017-07-17 00:28:33             "removes": null, 
2017-07-17 00:28:33             "responses": {
2017-07-17 00:28:33                 "foo": "bar"
2017-07-17 00:28:33             }, 
2017-07-17 00:28:33             "timeout": 30
2017-07-17 00:28:33         }
2017-07-17 00:28:33     }, 
2017-07-17 00:28:33     "rc": 0, 
2017-07-17 00:28:33     "start": "2017-07-16 20:28:32.932191", 
2017-07-17 00:28:33     "stdout": "", 
2017-07-17 00:28:33     "stdout_lines": []
2017-07-17 00:28:33 }
2017-07-17 00:28:33 
2017-07-17 00:28:33 TASK [expect : assert chdir works] *********************************************
2017-07-17 00:28:33 task path: /private/var/root/ansible/test/integration/targets/expect/tasks/main.yml:70
2017-07-17 00:28:33 fatal: [testhost]: FAILED! => {
2017-07-17 00:28:33     "assertion": "chdir_result.stdout == '/private/var/root/ansible_testing'", 
2017-07-17 00:28:33     "changed": false, 
2017-07-17 00:28:33     "evaluated_to": false, 
2017-07-17 00:28:33     "failed": true
2017-07-17 00:28:33 }
```
Close and reopen.

dagwieers commented:
@privateip @gundalow @caphrim007 Are these unit-tests sufficient to get this merged ? https://github.com/ansible/community/issues/110#issuecomment-314805189

Mind you that the JSON and XML error messages dependent heavily on resp. python version (2.6, 2.7 or 3+) and lxml version (3.2.1 and 3.8.0). Which makes unit-testing rather challenging.

caphrim007 commented on test/units/modules/network/aci/test_aci_rest.py:
```diff
@@ -0,0 +1,299 @@
+    def test_invalid_aci_login(self):
+        self.maxDiff = None
+
+        expected_result = {
```
my personal preference is to stuff things like this into a fixture file and then load them using something like load_json method that I found (at one time, dunno if it's there anymore) in the nxos module unit tests.

Here's what it looked like at the time
https://github.com/ansible/ansible/blob/devel/test/units/modules/network/f5/test_bigip_hostname.py#L58

and then stuff the JSON in a `fixtures/` directory

caphrim007 commented on test/units/modules/network/aci/test_aci_rest.py:
```diff
@@ -0,0 +1,299 @@
+            u'totalCount': '1',
+        }
+
+        json_response = '{"totalCount":"1","imdata":[{"error":{"attributes":{"code":"401","text":"Username or password is incorrect - FAILED local authentication"}}}]}'  # NOQA
```
another fixture here with load_json

caphrim007 commented on test/units/modules/network/aci/test_aci_rest.py:
```diff
@@ -0,0 +1,299 @@
+        if sys.version_info < (2, 7):
+            return
+
+        xml_response = '''<?xml version="1.0" encoding="UTF-8"?><imdata totalCount="1">
```
fixture with load_json. If it's not valid json then load_json just gives you back a plain string

caphrim007 commented on test/units/modules/network/aci/test_aci_rest.py:
```diff
@@ -0,0 +1,299 @@
+    def test_valid_aci_login(self):
+        self.maxDiff = None
+
+        expected_result = {
```
fixture with load_json

caphrim007 commented on test/units/modules/network/aci/test_aci_rest.py:
```diff
@@ -0,0 +1,299 @@
+            u'totalCount': u'1',
+        }
+
+        json_response = '{"totalCount":"1","imdata":[{"aaaLogin":{"attributes":{"token":"ZldYAsoO9d0FfAQM8xaEVWvQPSOYwpnqzhwpIC1r4MaToknJjlIuAt9+TvXqrZ8lWYIGPj6VnZkWiS8nJfaiaX/AyrdD35jsSxiP3zydh+849xym7ALCw/fFNsc7b5ik1HaMuSUtdrN8fmCEUy7Pq/QNpGEqkE8m7HaxAuHpmvXgtdW1bA+KKJu2zY1c/tem","siteFingerprint":"NdxD72K/uXaUK0wn","refreshTimeoutSeconds":"600","maximumLifetimeSeconds":"86400","guiIdleTimeoutS ...
```
fixture with load_json

caphrim007 commented on test/units/modules/network/aci/test_aci_rest.py:
```diff
@@ -0,0 +1,299 @@
+        if sys.version_info < (2, 7):
+            return
+
+        xml_response = '<?xml version="1.0" encoding="UTF-8"?><imdata totalCount="1">\n<aaaLogin token="ZldYAsoO9d0FfAQM8xaEVWvQPSOYwpnqzhwpIC1r4MaToknJjlIuAt9+TvXqrZ8lWYIGPj6VnZkWiS8nJfaiaX/AyrdD35jsSxiP3zydh+849xym7ALCw/fFNsc7b5ik1HaMuSUtdrN8fmCEUy7Pq/QNpGEqkE8m7HaxAuHpmvXgtdW1bA+KKJu2zY1c/tem" siteFingerprint="NdxD72K/uXaUK0wn" refreshTimeoutSeconds="600" maximumLifetimeSeconds="86400" guiIdle ...
```
fixture with  load_json

caphrim007 commented on test/units/modules/network/aci/test_aci_rest.py:
```diff
@@ -0,0 +1,299 @@
+    def test_invalid_input(self):
+        self.maxDiff = None
+
+        expected_result = {
```
fixture

caphrim007 commented on test/units/modules/network/aci/test_aci_rest.py:
```diff
@@ -0,0 +1,299 @@
+            u'totalCount': '1',
+        }
+
+        json_response = '{"totalCount":"1","imdata":[{"error":{"attributes":{"code":"401","text":"Username or password is incorrect - FAILED local authentication"}}}]}'  # NOQA
```
fixture

caphrim007 commented on test/units/modules/network/aci/test_aci_rest.py:
```diff
@@ -0,0 +1,299 @@
+        if sys.version_info < (2, 7):
+            return
+
+        xml_response = '''<?xml version="1.0" encoding="UTF-8"?><imdata totalCount="1">
```
fixture

caphrim007 commented on test/units/modules/network/aci/test_aci_rest.py:
```diff
@@ -0,0 +1,299 @@
+        self.maxDiffi = None
+
+        if sys.version_info < (3, 0):
+            expected_json_result = {
```
fixture

caphrim007 commented on test/units/modules/network/aci/test_aci_rest.py:
```diff
@@ -0,0 +1,299 @@
+            }
+
+        else:
+            expected_json_result = {
```
fixture

caphrim007 commented on test/units/modules/network/aci/test_aci_rest.py:
```diff
@@ -0,0 +1,299 @@
+            return
+
+        elif etree.LXML_VERSION < (3, 3, 0, 0):
+            expected_xml_result = {
```
fixture

caphrim007 commented on test/units/modules/network/aci/test_aci_rest.py:
```diff
@@ -0,0 +1,299 @@
+            }
+
+        elif sys.version_info < (3, 0):
+            expected_xml_result = {
```
fixture

caphrim007 commented on test/units/modules/network/aci/test_aci_rest.py:
```diff
@@ -0,0 +1,299 @@
+            }
+
+        else:
+            expected_xml_result = {
```
fixture

caphrim007 commented on test/units/modules/network/aci/test_aci_rest.py:
```diff
@@ -0,0 +1,299 @@
+        self.maxDiff = None
+
+        if sys.version_info < (2, 7):
+            expected_json_result = {
```
fixture

caphrim007 commented on test/units/modules/network/aci/test_aci_rest.py:
```diff
@@ -0,0 +1,299 @@
+            }
+
+        elif sys.version_info < (3, 0):
+            expected_json_result = {
```
fixture

caphrim007 commented on test/units/modules/network/aci/test_aci_rest.py:
```diff
@@ -0,0 +1,299 @@
+            }
+
+        else:
+            expected_json_result = {
```
fixture

caphrim007 commented on test/units/modules/network/aci/test_aci_rest.py:
```diff
@@ -0,0 +1,299 @@
+            return
+
+        elif etree.LXML_VERSION < (3, 3, 0, 0):
+            expected_xml_result = {
```
fixture

caphrim007 commented on test/units/modules/network/aci/test_aci_rest.py:
```diff
@@ -0,0 +1,299 @@
+            }
+
+        elif sys.version_info < (3, 0):
+            expected_xml_result = {
```
fixture

caphrim007 commented on test/units/modules/network/aci/test_aci_rest.py:
```diff
@@ -0,0 +1,299 @@
+            }
+
+        else:
+            expected_xml_result = {
```
fixture

caphrim007 commented on test/units/modules/network/aci/test_aci_rest.py:
```diff
@@ -0,0 +1,299 @@
+        self.assertEqual(expected_result, json_result)
+
+        # Python 2.7+ is needed for xmljson
+        if sys.version_info < (2, 7):
```
my personal preference here is to dump all of these 2.7 specific modules into a Test class that has a pytest decorator of @pytest.skipif(sys.version_info <(2,7)) or whatever the syntax is.

By collecting the version specific tests in their own classes, you can avoid this conditional logic in the test itself

dagwieers replied:
It's not that easy, we have different input whether it is python 2.6, or depending on the version of a library. So I have "if X, elseif Y, else" constructs that seem to be harder to do with decorators. (Or at least could lead into unsupported test-cases)

caphrim007 commented on test/units/modules/network/aci/test_aci_rest.py:
```diff
@@ -0,0 +1,299 @@
+    def test_empty_response(self):
+        self.maxDiffi = None
+
+        if sys.version_info < (3, 0):
```
Another test that I would move into version specific classes with pytest skipifs

caphrim007 commented on test/units/modules/network/aci/test_aci_rest.py:
```diff
@@ -0,0 +1,299 @@
+        self.assertEqual(expected_json_result, json_result)
+
+        # Python 2.7+ is needed for xmljson
+        if sys.version_info < (2, 7):
```
Another test that I would move into version specific classes with pytest skipifs

caphrim007 commented on test/units/modules/network/aci/test_aci_rest.py:
```diff
@@ -0,0 +1,299 @@
+        if sys.version_info < (2, 7):
+            return
+
+        elif etree.LXML_VERSION < (3, 3, 0, 0):
```
Another test that I would move into version specific classes with pytest skipifs

caphrim007 commented on test/units/modules/network/aci/test_aci_rest.py:
```diff
@@ -0,0 +1,299 @@
+                'raw': '',
+            }
+
+        elif sys.version_info < (3, 0):
```
Another test that I would move into version specific classes with pytest skipifs

caphrim007 commented on test/units/modules/network/aci/test_aci_rest.py:
```diff
@@ -0,0 +1,299 @@
+    def test_invalid_response(self):
+        self.maxDiff = None
+
+        if sys.version_info < (2, 7):
```
Another test that I would move into version specific classes with pytest skipifs

caphrim007 commented on test/units/modules/network/aci/test_aci_rest.py:
```diff
@@ -0,0 +1,299 @@
+                'raw': '{ "aaa":',
+            }
+
+        elif sys.version_info < (3, 0):
```
Another test that I would move into version specific classes with pytest skipifs

caphrim007 commented on test/units/modules/network/aci/test_aci_rest.py:
```diff
@@ -0,0 +1,299 @@
+        self.assertEqual(expected_json_result, json_result)
+
+        # Python 2.7+ is needed for xmljson
+        if sys.version_info < (2, 7):
```
Another test that I would move into version specific classes with pytest skipifs

caphrim007 commented on test/units/modules/network/aci/test_aci_rest.py:
```diff
@@ -0,0 +1,299 @@
+                'raw': '<aaa ',
+            }
+
+        elif sys.version_info < (3, 0):
```
Another test that I would move into version specific classes with pytest skipifs

caphrim007 reviewed (commented):
@dagwieers comments attached. Make of them what you will. Just observations that I've made over the course of writing my own tests. I'm flexible on the changes.

Maybe others can also chime in?

dagwieers commented:
@caphrim007 Thanks for the feedback, since this will be moved to module_utils, I am going to do this as part of that effort.

dagwieers commented:
+label new_module

ganeshrn commented on lib/ansible/modules/network/aci/aci_rest.py:
```diff
@@ -0,0 +1,395 @@
+        # NOTE: The XML-to-JSON conversion is using the "Cobra" convention
+        xmldata = None
+        try:
+            xml = lxml.etree.fromstring(to_bytes(rawoutput))
```
Can the native ElementTree api be used here.
```from xml.etree.ElementTree import fromstring```

This will reduce module dependency on `lxml`.

dagwieers replied:
We need lxml for [xmljson](https://github.com/sanand0/xmljson) anyway.
I think the main reason for xmljson to require lxml is the speed for conversion.

ganeshrn replied:
Ok, sounds good!
